File: pset6/sentimental-credit/credit.py
import re
import cs50


def main():
 # prompt for credit card number
    while True:
        try:
            card_number = cs50.get_string("Number: ")
            if re.match(r"^\d+$", card_number):  # Check if input contains only digits
                card_number = int(card_number)
                if card_number > 0:
                    break
                else:
                    print("Please enter a positive number.")
            else:
                print("Invalid input. Please enter digits only.")
        except ValueError:
            print("Invalid input. Please enter a valid number.")

    # validate the card number using the Luhn algorithm
    if not validity(card_number):
        print("INVALID")
        return

    # get card type
    card_type = get_card_type(card_number)

    # print card type as required
    print(card_type)


def validity(number):

    # validate credit card numbers using Luhn's algorithm
    num_str = str(number)
    sum = 0
    length = len(num_str)
    is_second_digit = False

    for i in range(length - 1, -1, -1):
        digit = int(num_str[i])

        if is_second_digit:
            digit *= 2

            if digit > 9:
                digit = digit % 10 + digit // 10

        sum += digit
        is_second_digit = not is_second_digit

    return sum % 10 == 0


def get_card_type(number):
    # determine card type based on number patterns
    num_str = str(number)
    length = len(num_str)

    # check for American Express(AMEX)
    if re.match(r"^(34|37)", num_str) and length == 15:
        return "AMEX"

    # chexk for Mastercard
    if re.match(r"^(51|52|53|54|55)", num_str) and length == 16:
        return "MASTERCARD"

    # chexk for Visa
    if re.match(r"^4", num_str) and (length == 13 or length == 16):
        return "VISA"

    return "INVALID"


main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool validity(long long number);
string get_card_type(long long number);

int main(void)
{
    // Prompt user for credit card number
    long long card_number;
    do
    {
        card_number = get_long_long("Number: ");
    }
    while (card_number < 0);

    // Validate the card number using the luhn algorithm
    if (!validity(card_number))
    {
        printf("INVALID\n");
        return 0;
    }

    // Get the card type
    string card_type = get_card_type(card_number);

    // Print the card type
    printf("%s\n", card_type);

    return 0;
}

// Function to validate credit card number
bool validity(long long number)
{
    int sum = 0;
    bool double_digit = false;

    while (number > 0)
    {
        int digit = number % 10;

        if (double_digit)
        {
            digit *= 2;

            // If doubling results in a two digit number, sum the digits of that number
            if (digit >= 10)
            {
                sum += (digit % 10) + (digit / 10);
            }
            else
            {
                sum += digit;
            }
        }
        else
        {
            sum += digit;
        }

        // Toggle the flag for next digit
        double_digit = !double_digit;

        // Remove the last digit from the number
        number /= 10;
    }

    // Check if the sum modulo 10 equals 0
    return (sum % 10 == 0);
}

// Function to determine card type based on first digits of the card number
string get_card_type(long long number)
{
    // Check for american express
    if ((number >= 340000000000000 && number < 350000000000000) ||
        (number >= 370000000000000 && number < 380000000000000))
    {
        return "AMEX";
    }

    // Check for mastercard
    if (number >= 5100000000000000 && number < 5600000000000000)
    {
        return "MASTERCARD";
    }

    // Check for visa
    if ((number >= 4000000000000 && number < 5000000000000) ||
        (number >= 4000000000000000 && number < 5000000000000000))
    {
        return "VISA";
    }

    // Default case if invalid
    return "INVALID";
}
